Could not install emacs@30

<!-- To help us debug your issue, please complete these sections: -->

Please make sure to follow these steps (and mark the checkboxes):

- [x] run `brew update` and try to reproduce the issue again
- [x] run `brew doctor`, fix all issues and try to reproduce your issue again
- [x] run `brew config` and `brew doctor` and include their output

## What you were trying to do

Install emacs 30 

## What happened (include command output)

<pre>
$ brew install emacs-plus@30 --with-native-comp
</pre>

<details>
  <summary>Command output</summary>
  <pre>
❯ brew install emacs-plus@30 --with-native-comp
==> Fetching d12frosted/emacs-plus/emacs-plus@30
==> Cloning https://github.com/emacs-mirror/emacs.git
Updating /Users/p/Library/Caches/Homebrew/emacs-plus@30--git
From https://github.com/emacs-mirror/emacs
   1c8b6a00cff..9c86dd52475  master     -> origin/master
==> Checking out branch master
Already on 'master'
Your branch is behind 'origin/master' by 408 commits, and can be fast-forwarded.
  (use "git pull" to update your local branch)
HEAD is now at 9c86dd52475 Tramp's direct asynchronous processes use 'tramp-remote-path'
==> Reinstalling d12frosted/emacs-plus/emacs-plus@30 --with-native-comp
==> Patching
==> Applying fix-window-role.patch
patching file src/nsterm.m
Hunk #1 succeeded at 9456 (offset 688 lines).
==> Applying system-appearance.patch
patching file src/frame.h
patching file src/nsfns.m
Hunk #1 succeeded at 1428 (offset 172 lines).
patching file src/nsterm.m
Hunk #1 succeeded at 1935 (offset 46 lines).
Hunk #2 succeeded at 5780 (offset 385 lines).
Hunk #3 succeeded at 6068 (offset 427 lines).
Hunk #4 succeeded at 6118 (offset 427 lines).
Hunk #5 succeeded at 6352 (offset 435 lines).
Hunk #6 succeeded at 9789 (offset 887 lines).
Hunk #7 succeeded at 11122 (offset 1064 lines).
==> Applying round-undecorated-frame.patch
patching file src/frame.c
Hunk #1 succeeded at 972 (offset 2 lines).
Hunk #2 succeeded at 4043 (offset 133 lines).
Hunk #3 succeeded at 6275 (offset 221 lines).
patching file src/frame.h
Hunk #1 succeeded at 459 (offset 63 lines).
Hunk #2 succeeded at 1239 (offset 157 lines).
Hunk #3 succeeded at 1266 (offset 157 lines).
patching file src/nsfns.m
Hunk #1 succeeded at 1100 with fuzz 2 (offset 51 lines).
Hunk #2 succeeded at 1428 with fuzz 2 (offset 51 lines).
patching file src/nsterm.h
Hunk #1 succeeded at 1238 (offset 11 lines).
patching file src/nsterm.m
Hunk #1 succeeded at 1801 (offset 3 lines).
Hunk #2 succeeded at 9226 (offset 149 lines).
Hunk #3 succeeded at 9314 (offset 149 lines).
==> ./autogen.sh
==> ./configure --disable-silent-rules --enable-locallisppath=/usr/local/share/emacs/site-lisp
==> gmake
Last 15 lines from /Users/p/Library/Logs/Homebrew/emacs-plus@30/03.gmake:
gmake[2]: Leaving directory '/private/tmp/emacs-plusA30-20231222-21283-onsvu4/lisp'
gmake[1]: *** [Makefile:537: lisp] Error 2
gmake[1]: Leaving directory '/private/tmp/emacs-plusA30-20231222-21283-onsvu4'
gmake[1]: Entering directory '/private/tmp/emacs-plusA30-20231222-21283-onsvu4'
***
*** "make all" failed with exit status 2.
***
*** You could try to:
*** - run "make bootstrap", which might fix the problem
*** - run "make V=1", which displays the full commands invoked by make,
***   to further investigate the problem
***
gmake[1]: *** [Makefile:418: advice-on-failure] Error 2
gmake[1]: Leaving directory '/private/tmp/emacs-plusA30-20231222-21283-onsvu4'
gmake: *** [Makefile:374: all] Error 2
  </pre>
</details>

## Output of `brew config`

<pre>
$ brew config
❯ brew config
HOMEBREW_VERSION: 4.2.0
ORIGIN: https://github.com/Homebrew/brew
HEAD: ff899e5b3ff6731208564ef502db237e2627e8ba
Last commit: 4 days ago
Core tap JSON: 21 Dec 17:36 UTC
HOMEBREW_PREFIX: /usr/local
HOMEBREW_BAT_THEME: gruvbox-light
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: nvim
HOMEBREW_MAKE_JOBS: 16
Homebrew Ruby: 3.1.4 => /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: 16-core 64-bit kabylake
Clang: 15.0.0 build 1500
Git: 2.43.0 => /usr/local/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-x86_64
CLT: 15.1.0.0.1.1700200546
Xcode: 15.1
</pre>


## Output of `brew doctor`

<pre>
$ brew doctor
❯ brew doctor
Please note that these warnings are just used to help the Homebrew maintainers
with debugging if you file an issue. If everything you use Homebrew for is
working fine: please don't worry or file an issue; just ignore this. Thanks!

Warning: Some installed casks are deprecated or disabled.
You should find replacements for the following casks:
  ksdiff

Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  libav

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  dog
  exa
  gcal
  libav
  mcrypt
  openssl@1.1
  youtube-dl

Warning: Unbrewed dylibs were found in /usr/local/lib.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected dylibs:
  /usr/local/lib/libbasekit.dylib
  /usr/local/lib/libcoroutine.dylib
  /usr/local/lib/libgarbagecollector.dylib
  /usr/local/lib/libiovmall.dylib
  /usr/local/lib/libsample.dylib

Warning: Unbrewed header files were found in /usr/local/include.
If you didn't put them there on purpose they could cause problems when
building Homebrew formulae, and may need to be deleted.

Unexpected header files:
  /usr/local/include/io/386-ucontext.h
  /usr/local/include/io/BStream.h
  /usr/local/include/io/BStreamTag.h
  /usr/local/include/io/Base.h
  /usr/local/include/io/CHash.h
  /usr/local/include/io/CHash_inline.h
  /usr/local/include/io/Collector.h
  /usr/local/include/io/CollectorMarker.h
  /usr/local/include/io/CollectorMarker_inline.h
  /usr/local/include/io/Collector_inline.h
  /usr/local/include/io/Common.h
  /usr/local/include/io/Common_inline.h
  /usr/local/include/io/Coro.h
  /usr/local/include/io/Date.h
  /usr/local/include/io/Duration.h
  /usr/local/include/io/DynLib.h
  /usr/local/include/io/Hash_fnv.h
  /usr/local/include/io/Hash_murmur.h
  /usr/local/include/io/Hash_superfast.h
  /usr/local/include/io/IoBlock.h
  /usr/local/include/io/IoCFunction.h
  /usr/local/include/io/IoCall.h
  /usr/local/include/io/IoCollector.h
  /usr/local/include/io/IoCompiler.h
  /usr/local/include/io/IoConfig.h
  /usr/local/include/io/IoContext.h
  /usr/local/include/io/IoCoroutine.h
  /usr/local/include/io/IoDate.h
  /usr/local/include/io/IoDebugger.h
  /usr/local/include/io/IoDirectory.h
  /usr/local/include/io/IoDuration.h
  /usr/local/include/io/IoDynLib.h
  /usr/local/include/io/IoError.h
  /usr/local/include/io/IoFile.h
  /usr/local/include/io/IoFile_stat.h
  /usr/local/include/io/IoInstallPrefix.h
  /usr/local/include/io/IoLexer.h
  /usr/local/include/io/IoList.h
  /usr/local/include/io/IoMap.h
  /usr/local/include/io/IoMessage.h
  /usr/local/include/io/IoMessage_inline.h
  /usr/local/include/io/IoMessage_opShuffle.h
  /usr/local/include/io/IoMessage_parser.h
  /usr/local/include/io/IoNumber.h
  /usr/local/include/io/IoObject.h
  /usr/local/include/io/IoObject_flow.h
  /usr/local/include/io/IoObject_inline.h
  /usr/local/include/io/IoObject_struct.h
  /usr/local/include/io/IoProfiler.h
  /usr/local/include/io/IoSandbox.h
  /usr/local/include/io/IoSeq.h
  /usr/local/include/io/IoSeq_immutable.h
  /usr/local/include/io/IoSeq_inline.h
  /usr/local/include/io/IoSeq_mutable.h
  /usr/local/include/io/IoSeq_vector.h
  /usr/local/include/io/IoState.h
  /usr/local/include/io/IoState_callbacks.h
  /usr/local/include/io/IoState_coros.h
  /usr/local/include/io/IoState_debug.h
  /usr/local/include/io/IoState_eval.h
  /usr/local/include/io/IoState_exceptions.h
  /usr/local/include/io/IoState_inline.h
  /usr/local/include/io/IoState_symbols.h
  /usr/local/include/io/IoSystem.h
  /usr/local/include/io/IoTag.h
  /usr/local/include/io/IoTag_inline.h
  /usr/local/include/io/IoToken.h
  /usr/local/include/io/IoVM.h
  /usr/local/include/io/IoVMApi.h
  /usr/local/include/io/IoVersion.h
  /usr/local/include/io/IoWeakLink.h
  /usr/local/include/io/List.h
  /usr/local/include/io/List_inline.h
  /usr/local/include/io/MainArgs.h
  /usr/local/include/io/PHash.h
  /usr/local/include/io/PHash_inline.h
  /usr/local/include/io/PHash_struct.h
  /usr/local/include/io/PointerHash.h
  /usr/local/include/io/PointerHash_inline.h
  /usr/local/include/io/PointerHash_struct.h
  /usr/local/include/io/PortableGettimeofday.h
  /usr/local/include/io/PortableSorting.h
  /usr/local/include/io/PortableStdint.h
  /usr/local/include/io/PortableStrlcpy.h
  /usr/local/include/io/PortableStrptime.h
  /usr/local/include/io/PortableTruncate.h
  /usr/local/include/io/PortableUsleep.h
  /usr/local/include/io/RandomGen.h
  /usr/local/include/io/Stack.h
  /usr/local/include/io/Stack_inline.h
  /usr/local/include/io/UArray.h
  /usr/local/include/io/UArray_character.h
  /usr/local/include/io/UArray_format.h
  /usr/local/include/io/UArray_math.h
  /usr/local/include/io/UArray_path.h
  /usr/local/include/io/UArray_stream.h
  /usr/local/include/io/UArray_string.h
  /usr/local/include/io/UArray_utf.h
  /usr/local/include/io/amd64-ucontext.h
  /usr/local/include/io/cdecode.h
  /usr/local/include/io/cencode.h
  /usr/local/include/io/power-ucontext.h
  /usr/local/include/io/taskimpl.h
  /usr/local/include/io/utf8.h
  /usr/local/include/io/utf8internal.h
  /usr/local/include/io/utf_convert.h
  /usr/local/include/python3.9/greenlet/greenlet.h

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python-certifi
  python-packaging
  python-click
  pygments
  pycparser
  cffi
  python-platformdirs
  python-filelock
  pyyaml
  python-typing-extensions
  python-cryptography
  six

Warning: Some installed formulae are not readable:
  clojure-lsp-native: /usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:4: syntax error, unexpected <<, expecting `end'
<<<<<<< HEAD
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:8: syntax error, unexpected '|', expecting `end'
||||||| parent of 420687c (Bum...
^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:8: syntax error, unexpected local variable or method, expecting `end'
||||||| parent of 420687c (Bump to version: 2022.01.20...
                        ^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:8: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:8: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:10: syntax error, unexpected ===, expecting `end'
=======
^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:12: syntax error, unexpected >>, expecting `end'
>>>>>>> 420687c (Bump to versio...
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:12: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:12: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:15: syntax error, unexpected <<
<<<<<<< HEAD
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:23: syntax error, unexpected '|', expecting `end'
||||||| parent of 420687c (Bum...
^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:23: syntax error, unexpected local variable or method, expecting `end'
||||||| parent of 420687c (Bump to version: 2022.01.20...
                        ^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:23: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:23: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:26: syntax error, unexpected ===, expecting `end'
=======
^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:29: syntax error, unexpected >>, expecting `end'
>>>>>>> 420687c (Bump to versio...
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:29: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:29: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:31: syntax error, unexpected <<
<<<<<<< HEAD
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:39: syntax error, unexpected '|', expecting `end'
||||||| parent of 420687c (Bum...
^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:39: syntax error, unexpected local variable or method, expecting `end'
||||||| parent of 420687c (Bump to version: 2022.01.20...
                        ^
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:39: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:39: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:42: syntax error, unexpected ===, expecting `end'
=======
^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:45: syntax error, unexpected >>, expecting `end'
>>>>>>> 420687c (Bump to versio...
^~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:45: unexpected fraction part after numeric literal
...7c (Bump to version: 2022.01.20-14.12.43)
...                            ^~~
/usr/local/Homebrew/Library/Taps/clojure-lsp/homebrew-brew/clojure-lsp-native.rb:45: unexpected fraction part after numeric literal
...to version: 2022.01.20-14.12.43)
...                            ^~~


Warning: Some taps are not on the default git origin branch and may not receive
updates. If this is a surprise to you, check out the default branch with:
  git -C $(brew --repo clojure-lsp/brew) checkout main

Warning: Homebrew's "sbin" was not found in your PATH but you have installed
formulae that put executables in /usr/local/sbin.
Consider setting your PATH for example like so:
  fish_add_path /usr/local/sbin

Warning: Some installed formulae are missing dependencies.
You should `brew install` the missing dependencies:
  brew install sdl12-compat

Run `brew missing` for more details.
</pre>

--- a/Formula/emacs-plus@30.rb
+++ b/Formula/emacs-plus@30.rb
@@ -88,7 +88,7 @@ class EmacsPlusAT30 < EmacsBase
 
   opoo "The option --with-no-frame-refocus is not required anymore in emacs-plus@30." if build.with? "no-frame-refocus"
   local_patch "fix-window-role", sha: "1f8423ea7e6e66c9ac6dd8e37b119972daa1264de00172a24a79a710efcb8130"
-  local_patch "system-appearance", sha: "d6ee159839b38b6af539d7b9bdff231263e451c1fd42eec0d125318c9db8cd92"
+  local_patch "system-appearance", sha: "9eb3ce80640025bff96ebaeb5893430116368d6349f4eb0cb4ef8b3d58477db6"
   local_patch "poll", sha: "052eacac5b7bd86b466f9a3d18bff9357f2b97517f463a09e4c51255bdb14648" if build.with? "poll"
   local_patch "round-undecorated-frame", sha: "7451f80f559840e54e6a052e55d1100778abc55f98f1d0c038a24e25773f2874"
 

--- a/patches/emacs-30/system-appearance.patch
+++ b/patches/emacs-30/system-appearance.patch
@@ -1,1 +1,316 @@
-../emacs-28/system-appearance.patch
+Patch to make emacs 28 aware of the macOS 10.14+ system appearance changes.
+
+From 6e73cd55ebfd3b0967357b3c3ead16d2f8539526 Mon Sep 17 00:00:00 2001
+From: "Nicolas G. Querol" <nicolas.gquerol@gmail.com>
+Date: Wed, 11 Nov 2020 12:35:47 +0100
+Subject: [PATCH] Add `ns-system-appearance-change-functions' hook
+
+This implements a new hook, effective only on macOS >= 10.14 (Mojave),
+that is called when the system changes its appearance (e.g. from light
+to dark). Users can then implement functions that take this change
+into account, for instance to load a particular theme.
+
+Minor changes are also made to select the right "dark" appearance
+(NSAppearanceNameDarkAqua) on macOS versions >= 10.14, the previous one
+(NSAppearanceNameVibrantDark) being deprecated.
+
+* src/frame.h (enum ns_appearance_type): Add new
+"ns_appearance_dark_aqua" case.
+
+* src/nsfns.m (defun x-create-frame): Use "dark aqua" appearance on
+macOS >= 10.14.
+
+* src/nsterm.m:
+  - (ns_set_appearance): Use "dark aqua" appearance on
+     macOS >= 10.14, reset appearance to the system one
+     if `ns-appearance' frame parameter is not set to
+     either `dark' or `light'.
+  - (initFrameFromEmacs): Use "dark aqua" appearance on
+     macOS >= 10.14.
+  - (EmacsApp) Add the `systemDidChangeAppearance' private method,
+    as well as the appropriate Key-Value Observing calls to update
+    the frame's appearance when the system (and thus the app's)
+    appearance changes.
+  - Add `ns-system-appearance-change-functions' hook variable and
+    symbol, to allow users to add functions that react to the
+    change of the system's appearance.
+  - Add `ns-system-appearance' variable, to allow users to consult
+    the current system appearance.
+
+Here is an example on how to use this new feature:
+
+    (defun my/load-theme (appearance)
+      "Load theme, taking current system APPEARANCE into consideration."
+      (mapc #'disable-theme custom-enabled-themes)
+      (pcase appearance
+        ('light (load-theme 'tango t))
+        ('dark (load-theme 'tango-dark t))))
+
+    (add-hook 'ns-system-appearance-change-functions #'my/load-theme)
+
+The hook being run on each system appearance change as well as at
+startup time, Emacs should then always load the appropriate theme.
+---
+ src/frame.h  |   3 +-
+ src/nsfns.m  |  13 ++++-
+ src/nsterm.m | 153 ++++++++++++++++++++++++++++++++++++++++++++++-----
+ 3 files changed, 153 insertions(+), 16 deletions(-)
+
+diff --git a/src/frame.h b/src/frame.h
+index a8ad011889..e7f7fdafe1 100644
+--- a/src/frame.h
++++ b/src/frame.h
+@@ -71,7 +71,8 @@ #define EMACS_FRAME_H
+   {
+     ns_appearance_system_default,
+     ns_appearance_aqua,
+-    ns_appearance_vibrant_dark
++    ns_appearance_vibrant_dark,
++    ns_appearance_dark_aqua
+   };
+ #endif
+ #endif /* HAVE_WINDOW_SYSTEM */
+diff --git a/src/nsfns.m b/src/nsfns.m
+index 07bcab1816..4766eb91ae 100644
+--- a/src/nsfns.m
++++ b/src/nsfns.m
+@@ -1256,14 +1256,25 @@ Turn the input menu (an NSMenu) into a lisp list for tracking on lisp side.
+   store_frame_param (f, Qundecorated, FRAME_UNDECORATED (f) ? Qt : Qnil);
+ 
+ #ifdef NS_IMPL_COCOA
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
+   tem = gui_display_get_arg (dpyinfo, parms, Qns_appearance, NULL, NULL,
+                              RES_TYPE_SYMBOL);
+   if (EQ (tem, Qdark))
+-    FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
++    if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14)
++      {
++        FRAME_NS_APPEARANCE (f) = ns_appearance_dark_aqua;
++      }
++    else
++      {
++        FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
++      }
+   else if (EQ (tem, Qlight))
+     FRAME_NS_APPEARANCE (f) = ns_appearance_aqua;
+   else
+     FRAME_NS_APPEARANCE (f) = ns_appearance_system_default;
++
+   store_frame_param (f, Qns_appearance,
+                      (!NILP (tem) && !EQ (tem, Qunbound)) ? tem : Qnil);
+ 
+diff --git a/src/nsterm.m b/src/nsterm.m
+index 4bdc67c10b..0d2f3e0b2b 100644
+--- a/src/nsterm.m
++++ b/src/nsterm.m
+@@ -1889,11 +1889,25 @@ Hide the window (X11 semantics)
+     return;
+ 
+   if (EQ (new_value, Qdark))
+-    FRAME_NS_APPEARANCE (f) = ns_appearance_vibrant_dark;
+-  else if (EQ (new_value, Qlight))
+-    FRAME_NS_APPEARANCE (f) = ns_appearance_aqua;
++    {
++#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
++      if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14)
++        FRAME_NS_APPEARANCE(f) = ns_appearance_dark_aqua;
++      else
++#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
++        FRAME_NS_APPEARANCE(f) = ns_appearance_vibrant_dark;
++    }
++  else if (EQ(new_value, Qlight))
++    {
++      FRAME_NS_APPEARANCE (f) = ns_appearance_aqua;
++    }
+   else
+-    FRAME_NS_APPEARANCE (f) = ns_appearance_system_default;
++    {
++      FRAME_NS_APPEARANCE (f) = ns_appearance_system_default;
++    }
+ 
+   [window setAppearance];
+ #endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101000 */
+@@ -5381,6 +5395,7 @@ Needs to be here because ns_initialize_display_info () uses AppKit classes.
+ 
+    ========================================================================== */
+ 
++static const void *kEmacsAppKVOContext = &kEmacsAppKVOContext;
+ 
+ @implementation EmacsApp
+ 
+@@ -5626,6 +5641,18 @@ - (void)applicationDidFinishLaunching: (NSNotification *)notification
+ 	 object:nil];
+ #endif
+ 
++#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++  [self addObserver:self
++         forKeyPath:NSStringFromSelector(@selector(effectiveAppearance))
++            options:NSKeyValueObservingOptionInitial|NSKeyValueObservingOptionNew
++            context:&kEmacsAppKVOContext];
++
++   pending_funcalls = Fcons(list3(Qrun_hook_with_args,
++                                  Qns_system_appearance_change_functions,
++                                  Vns_system_appearance),
++                            pending_funcalls);
++#endif
++
+ #ifdef NS_IMPL_COCOA
+   /* Some functions/methods in CoreFoundation/Foundation increase the
+      maximum number of open files for the process in their first call.
+@@ -5664,6 +5691,68 @@ - (void)antialiasThresholdDidChange:(NSNotification *)notification
+ #endif
+ }
+ 
++- (void)observeValueForKeyPath:(NSString *)keyPath
++                      ofObject:(id)object
++                        change:(NSDictionary *)change
++                       context:(void *)context
++{
++#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++  if (context == kEmacsAppKVOContext
++      && object == self
++      && [keyPath isEqualToString:
++                    NSStringFromSelector (@selector(effectiveAppearance))])
++    [self systemAppearanceDidChange:
++               [change objectForKey:NSKeyValueChangeNewKey]];
++  else
++#endif /* (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
++    [super observeValueForKeyPath:keyPath
++                         ofObject:object
++                           change:change
++                          context:context];
++}
++
++#if defined (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
++- (void)systemAppearanceDidChange:(NSAppearance *)newAppearance
++{
++
++  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_14)
++    return;
++
++  NSAppearanceName appearance_name =
++      [newAppearance bestMatchFromAppearancesWithNames:@[
++        NSAppearanceNameAqua, NSAppearanceNameDarkAqua
++      ]];
++
++  BOOL is_dark_appearance =
++    [appearance_name isEqualToString:NSAppearanceNameDarkAqua];
++  Vns_system_appearance = is_dark_appearance ? Qdark : Qlight;
++
++  run_system_appearance_change_hook ();
++}
++
++static inline void run_system_appearance_change_hook (void)
++{
++  if (NILP (Vns_system_appearance_change_functions))
++    return;
++
++  block_input ();
++
++  bool owfi = waiting_for_input;
++  waiting_for_input = false;
++
++  safe_calln (Qrun_hook_with_args,
++          Qns_system_appearance_change_functions,
++          Vns_system_appearance);
++  Fredisplay(Qt);
++
++  waiting_for_input = owfi;
++
++  unblock_input ();
++}
++#endif /* (NS_IMPL_COCOA) && MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
+ 
+ /* Termination sequences:
+     C-x C-c:
+@@ -5828,6 +5917,14 @@ - (void)applicationDidResignActive: (NSNotification *)notification
+   ns_send_appdefined (-1);
+ }
+ 
++- (void)applicationWillTerminate:(NSNotification *)notification
++{
++  NSTRACE ("[EmacsApp applicationWillTerminate:]");
++
++  [self removeObserver:self
++            forKeyPath:NSStringFromSelector(@selector(effectiveAppearance))
++               context:&kEmacsAppKVOContext];
++}
+ 
+ 
+ /* ==========================================================================
+@@ -8805,17 +8902,26 @@ - (void)setAppearance
+ #define NSAppKitVersionNumber10_10 1343
+ #endif
+ 
+-  if (NSAppKitVersionNumber < NSAppKitVersionNumber10_10)
+-    return;
+-
+-  if (FRAME_NS_APPEARANCE (f) == ns_appearance_vibrant_dark)
+-    appearance =
+-      [NSAppearance appearanceNamed:NSAppearanceNameVibrantDark];
+-  else if (FRAME_NS_APPEARANCE (f) == ns_appearance_aqua)
+-    appearance =
+-      [NSAppearance appearanceNamed:NSAppearanceNameAqua];
++   if (NSAppKitVersionNumber < NSAppKitVersionNumber10_10)
++     return;
+ 
+-  [self setAppearance:appearance];
++#if MAC_OS_X_VERSION_MAX_ALLOWED >= 101400
++#ifndef NSAppKitVersionNumber10_14
++#define NSAppKitVersionNumber10_14 1671
++#endif
++   if (NSAppKitVersionNumber >= NSAppKitVersionNumber10_14
++       && FRAME_NS_APPEARANCE(f) == ns_appearance_dark_aqua)
++     appearance = [NSAppearance appearanceNamed:NSAppearanceNameDarkAqua];
++   else
++#endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101400 */
++     if (FRAME_NS_APPEARANCE(f) == ns_appearance_vibrant_dark)
++       appearance =
++         [NSAppearance appearanceNamed:NSAppearanceNameVibrantDark];
++     else if (FRAME_NS_APPEARANCE (f) == ns_appearance_aqua)
++       appearance =
++         [NSAppearance appearanceNamed:NSAppearanceNameAqua];
++
++   [self setAppearance:appearance];
+ #endif /* MAC_OS_X_VERSION_MAX_ALLOWED >= 101000 */
+ }
+ 
+@@ -9952,6 +10058,25 @@ Nil means use fullscreen the old (< 10.7) way.  The old way works better with
+ This variable is ignored on macOS < 10.7 and GNUstep.  Default is t.  */);
+   ns_use_mwheel_momentum = YES;
+ 
++ DEFVAR_LISP ("ns-system-appearance", Vns_system_appearance,
++               doc: /* Current system appearance, i.e. `dark' or `light'.
++
++This variable is ignored on macOS < 10.14 and GNUstep.  Default is nil.  */);
++  Vns_system_appearance = Qnil;
++  DEFSYM(Qns_system_appearance, "ns-system-appearance");
++
++  DEFVAR_LISP ("ns-system-appearance-change-functions",
++               Vns_system_appearance_change_functions,
++     doc: /* List of functions to call when the system appearance changes.
++Each function is called with a single argument, which corresponds to the new
++system appearance (`dark' or `light').
++
++This hook is also run once at startup.
++
++This variable is ignored on macOS < 10.14 and GNUstep.  Default is nil.  */);
++  Vns_system_appearance_change_functions = Qnil;
++  DEFSYM(Qns_system_appearance_change_functions, "ns-system-appearance-change-functions");
++
+   /* TODO: Move to common code.  */
+   DEFVAR_LISP ("x-toolkit-scroll-bars", Vx_toolkit_scroll_bars,
+ 	       doc: /* SKIP: real doc in xterm.c.  */);
+
+base-commit: e5c481b61c26bcf83779db9fb3ac6b96bc50ab2e
+-- 
+2.33.0
+

